src/main/java/com/cloudbees/jenkins/support/threaddump/HighLoadChecker.java

```diff
@@ -0,0 +1,116 @@
+import hudson.Extension;
+import hudson.ExtensionList;
+import hudson.model.PeriodicWork;
+import jenkins.metrics.api.MetricProvider;
+import jenkins.metrics.impl.VMMetricProviderImpl;
+import jenkins.model.Jenkins;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.WARNING;
+
+/**
+ * PeriodicWork to check when there is a high load in the instance.
+ *
+ * Only checking high CPU usage for the moment, but it can be used
+ * to generate thread dumps in high heap memory consumption.
+ */
+@Extension
+public class HighLoadChecker extends PeriodicWork {
+
+    /**
+     * Recurrence period to check high load consumption and take the corresponded actions
+     * like generating a thread dump
+     */
+    public static final int RECURRENCE_PERIOD_SEC =
+            Integer.getInteger(HighLoadChecker.class.getName()+ ".RECURRENCE_PERIOD_SEC", 900);
+
+    /**
+     * This is the CPU usage threshold. Determinate de percentage of the total CPU used across all the
+     * cores available
+     */
+    public static final Double CPU_USAGE_THRESHOLD =
+            new Double(System.getProperty(HighLoadChecker.class.getName() + ".CPU_USAGE_THRESHOLD", "0.95"));
+
+    /**
+     * Limit the number of thread dumps to retain on high cpu
+     */
+    public static final int HIGH_CPU_THEAD_DUMPS_TO_RETAIN =
+            Integer.getInteger(HighLoadChecker.class.getName()+ ".HIGH_CPU_THEAD_DUMPS_TO_RETAIN", 10);
+
+    /**
+     * Thread dumps generated on high CPU load are stored in $JENKINS_HOME/high-load/cpu
+     **/
+    final FileListCap logs = new FileListCap(new File(Jenkins.getInstance().getRootDir(),"high-load/cpu"), HIGH_CPU_THEAD_DUMPS_TO_RETAIN);
+
+    final SimpleDateFormat format = new SimpleDateFormat("yyyyMMdd-HHmmss.SSS");
+    {
+        format.setTimeZone(TimeZone.getTimeZone("UTC"));
+    }
+
+    @Override
+    public long getRecurrencePeriod() {
+        return TimeUnit.SECONDS.toMillis(RECURRENCE_PERIOD_SEC);
+    }
+
+    @Override
+    protected void doRun() throws Exception {
+        Double cpuLoad = null;
+        try {
+            VMMetricProviderImpl vmMetricProvider = ExtensionList.lookup(MetricProvider.class).get(VMMetricProviderImpl.class);
+            MetricSet vmProviderMetricSet = vmMetricProvider.getMetricSet();
+            Gauge cpu = (Gauge) vmProviderMetricSet.getMetrics().get("vm.cpu.load");
+            cpuLoad = new Double(cpu.getValue().toString());
+        } catch (NullPointerException nullPointerException) {
+            LOGGER.log(WARNING, "Support Core plugin can't generate thread dumps. Metrics plugin does not seem to be available", nullPointerException);
+            return;
+        }
+
+        if (Double.compare(Runtime.getRuntime().availableProcessors() * CPU_USAGE_THRESHOLD, cpuLoad) < 0) {
```

Comment:
@varyvol AFAIK this is correct, doesn't it? If `cpuLoad > Runtime.getRuntime().availableProcessors() * CPU_USAGE_THRESHOLD`, then  it will return 

```
     * @param   d1        the first {@code double} to compare
     * @param   d2        the second {@code double} to compare
     * @return  the value {@code 0} if {@code d1} is
     *          numerically equal to {@code d2}; a value less than
     *          {@code 0} if {@code d1} is numerically less than
     *          {@code d2}; and a value greater than {@code 0}
     *          if {@code d1} is numerically greater than
     *          {@code d2}.
     * @since 1.4
     */
```

Reply:
Yes, I believe it's valid but it's really confusing to me. At least my thinking is: I've calculated the cpuLoad, great! Now I'm going to compare it to the limit. Is it bigger? Then I do whatever.

But not against it if you think it's fine.

Reply:
Looks correct to me. 
Another thing I want to point out is that if the cpuLoad is `Double#NaN`, then it will satisfy this condition and generate a thread dump and we may not want this. I think this is a possibility and maybe we should check that the `cpuLoad` is not `NaN` to avoid false positive ?